exla/c_src/exla/exla_mlir.cc

```diff
@@ -0,0 +1,145 @@
+                                        builder->getStringAttr(op_name),
+                                        operands_range,
+                                        types_range,
+                                        attributes_array,
+                                        {},
+                                        {}};
+
+  for (auto region : regions) {
+    auto new_region = op_state.addRegion();
+    new_region->getBlocks().splice(new_region->end(), region->getBlocks());
+  }
+
+  auto op = builder->create(op_state);
+
+  auto results = op->getResults();
+  auto results_vec = std::vector<mlir::Value>(results.begin(), results.end());
+
+  return results_vec;
+}
+
+std::pair<mlir::Region *, std::vector<mlir::Value>> MLIRFunction::PushRegion(std::vector<mlir::Type> types) {
+  auto context = module_->builder()->getContext();
+
+  auto region = new mlir::Region();
+  auto & block = region->emplaceBlock();
+
+  for (mlir::Type type : types) {
+    block.addArgument(type, mlir::UnknownLoc::get(context));
+  }
+
+  auto args = std::vector<mlir::Value>{};
+  for (auto &arg : block.getArguments()) {
+    args.push_back(arg);
+  }
+
+  region_stack.push(std::move(region));
+  setInsertionPoint();
+
+  return {region, args};
+}
+
+void MLIRFunction::PopRegion() {
+  region_stack.pop();
+  setInsertionPoint();
+}
+
+void MLIRFunction::setInsertionPoint() {
+  if (region_stack.size() == 0) {
+    module_->builder()->setInsertionPointToEnd(&func_->getBody().back());
+  } else {
+    module_->builder()->setInsertionPointToEnd(&region_stack.top()->back());
+  }
+}
+
+MLIRModule::MLIRModule(mlir::MLIRContext *context) {
+  context_ = context;
+  module_ = mlir::OwningOpRef<mlir::ModuleOp>(mlir::ModuleOp::create(mlir::UnknownLoc::get(context_)));
+  builder_ = std::make_unique<mlir::OpBuilder>(context_);
+  builder_->setInsertionPointToStart(module_->getBody());
+}
+
+MLIRFunction *MLIRModule::CreateFunction(
+    std::string name,
+    std::vector<mlir::Type> arg_types,
+    std::vector<mlir::Type> ret_types,
+    bool is_public) {
+  auto visibility = is_public ? "public" : "nested";
+
+  auto funcType = builder_->getFunctionType(arg_types, ret_types);
+  auto loc = builder_->getUnknownLoc();
+  auto funcOp = std::make_unique<mlir::func::FuncOp>(mlir::func::FuncOp::create(loc, name, funcType));
+  funcOp->setSymVisibility(visibility);
+  module_->push_back(*funcOp);
+  funcOp->addEntryBlock();
+  builder_->setInsertionPointToStart(&funcOp->getBody().front());
+
+  return new MLIRFunction(this, std::move(funcOp));
+}
+
+std::string MLIRModule::ToString() {
+  auto output_string = std::string{};
+  auto output_stream = llvm::raw_string_ostream{output_string};
+  module_->print(output_stream);
+  return output_string;
+}
+
+mlir::Type MLIRModule::ParseType(std::string string) {
+  auto type = mlir::parseType(string, context_);
+
+  if (type == nullptr) {
+    std::cerr << "Unable to parse MLIR type: " << string << std::endl;
+    exit(1);
+  }
```

Comment:
Probably better to return this as an error instead of crashing the VM

Reply:
I did that initially, but then moved here to have the check in a single place. But yeah, I will revert. One detail is that `mlir::parseType` also emits a diagnostic, with possibly a more specific parsing error and I'm not sure if there's a way to capture that. That said, I will defer the error to outside NIF :)